### PR TITLE
[New Feature][Solace Integration-04] Portals related implementation to access Solace APIs.

### PR DESCRIPTION
**Pull request by Chamindu36**
### Purpose
This PR will add the REST API implementation to support accessing the Solace integration feature through UI portals.

### Goal
Fixes : 
https://github.com/wso2/product-apim/issues/11726

### Related PRs 
Marge after the following PR - 
https://github.com/wso2/carbon-apimgt/pull/10841

### CODE REVIEW

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/definitions/AsyncApiParser.java [hunk 1]
```diff
@@ -1888,4 +1889,20 @@ public String updateAsyncAPIDefinition(String oldDefinition, API apiToUpdate) {
...
         return wsUriMapping;
     }
+
+    public boolean isSolaceAPI(String definition) {
+        Aai20Document aai20Document = (Aai20Document) Library.readDocumentFromJSONString(definition);
+        Extension origin = aai20Document.info.getExtension("x-origin");
+        if (origin != null) {
+            ObjectMapper objectMapper = new ObjectMapper();
+            Map originMap = objectMapper.convertValue(origin.value, Map.class);
+            if (originMap.containsKey("vendor")) {
```
Comment (dnwick) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/definitions/AsyncApiParser.java:1899]: we can have a constant for vendor

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.admin.v1/src/gen/java/org/wso2/carbon/apimgt/rest/api/admin/v1/dto/AdditionalPropertyDTO.java [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+import com.fasterxml.jackson.annotation.JsonCreator;
+import io.swagger.annotations.ApiModel;
+import io.swagger.annotations.ApiModelProperty;
+import javax.validation.constraints.*;
+
+
+import io.swagger.annotations.*;
+import java.util.Objects;
+
+import javax.xml.bind.annotation.*;
```
Comment (dnwick) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.admin.v1/src/gen/java/org/wso2/carbon/apimgt/rest/api/admin/v1/dto/AdditionalPropertyDTO.java:13]: cant we import only relevant libraries without * ?

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.admin.v1/src/gen/java/org/wso2/carbon/apimgt/rest/api/admin/v1/dto/AdditionalPropertyDTO.java [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+    private String key = null;
+    private String value = null;
+
+    /**
+     **/
+    public AdditionalPropertyDTO key(String key) {
+        this.key = key;
+        return this;
+    }
+
```
Comment (dnwick) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.admin.v1/src/gen/java/org/wso2/carbon/apimgt/rest/api/admin/v1/dto/AdditionalPropertyDTO.java:32]: Fix formatting

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.admin.v1/src/gen/java/org/wso2/carbon/apimgt/rest/api/admin/v1/dto/AdditionalPropertyDTO.java [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+import javax.validation.Valid;
+
+
+
+public class AdditionalPropertyDTO   {
+
+    private String key = null;
+    private String value = null;
+
+    /**
```
Comment (dnwick) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.admin.v1/src/gen/java/org/wso2/carbon/apimgt/rest/api/admin/v1/dto/AdditionalPropertyDTO.java:26]: Add doc comments in proper format

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.admin.v1/src/gen/java/org/wso2/carbon/apimgt/rest/api/admin/v1/dto/AdditionalPropertyDTO.java [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+    @ApiModelProperty(example = "Organization", value = "")
+    @JsonProperty("key")
+    public String getKey() {
+        return key;
+    }
+    public void setKey(String key) {
+        this.key = key;
+    }
+
+    /**
```
Comment (dnwick) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.admin.v1/src/gen/java/org/wso2/carbon/apimgt/rest/api/admin/v1/dto/AdditionalPropertyDTO.java:43]: same here

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.admin.v1/src/gen/java/org/wso2/carbon/apimgt/rest/api/admin/v1/dto/AdditionalPropertyDTO.java [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+    public int hashCode() {
+        return Objects.hash(key, value);
+    }
+
+    @Override
+    public String toString() {
+        StringBuilder sb = new StringBuilder();
+        sb.append("class AdditionalPropertyDTO {\n");
+
+        sb.append("    key: ").append(toIndentedString(key)).append("\n");
```
Comment (dnwick) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.admin.v1/src/gen/java/org/wso2/carbon/apimgt/rest/api/admin/v1/dto/AdditionalPropertyDTO.java:84]: fix formatting please check all other places

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.store.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/store/v1/mappings/ApplicationMappingUtil.java [hunk 7]
```diff
@@ -62,6 +80,99 @@ public static ApplicationDTO fromApplicationtoDTO (Application application) {
...
+                                            }
+                                            applicationSolaceDeployedEnvironmentsDTO.setSolaceTopicsObject(solaceTopicsObjectDTO);
+                                        }
+                                    }
+                                }
+                                solaceEnvironments.add(applicationSolaceDeployedEnvironmentsDTO);
+                            }
+                        }
+                    } catch (IOException e) {
+                        e.printStackTrace();
```
Comment (dnwick) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.store.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/store/v1/mappings/ApplicationMappingUtil.java:167]: use loggers please

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.store.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/store/v1/mappings/ApplicationMappingUtil.java [hunk 8]
```diff
@@ -62,6 +80,99 @@ public static ApplicationDTO fromApplicationtoDTO (Application application) {
...
+                                        applicationSolaceDeployedEnvironmentsDTO.setOrganizationName(gatewayEnvironment.getAdditionalProperties().
+                                                get(APIConstants.SOLACE_ENVIRONMENT_ORGANIZATION));
+                                        boolean containsMQTTProtocol = false;
+                                        if (environmentObject.getJSONArray("messagingProtocols") != null) {
+                                            List<APISolaceURLsDTO> endpointUrls = new ArrayList<>();
+                                            JSONArray protocolsArray = environmentObject.getJSONArray("messagingProtocols");
+                                            for (int j = 0; j < protocolsArray.length(); j++) {
+                                                APISolaceURLsDTO solaceURLsDTO = new APISolaceURLsDTO();
+                                                String protocol = protocolsArray.getJSONObject(j).getJSONObject("protocol").getString("name");
+                                                if ("MQTT".equalsIgnoreCase(protocol)) {
```
Comment (dnwick) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.store.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/store/v1/mappings/ApplicationMappingUtil.java:135]: use a constant

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.store.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/store/v1/mappings/ApplicationMappingUtil.java (hunk 7 above)
Comment (dnwick) [reply components/apimgt/org.wso2.carbon.apimgt.rest.api.store.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/store/v1/mappings/ApplicationMappingUtil.java:167]: Check all other places and add more infor if can and log them

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.store.v1/src/gen/java/org/wso2/carbon/apimgt/rest/api/store/v1/dto/ApplicationSolaceDeployedEnvironmentsDTO.java [hunk 9]
```diff
@@ -0,0 +1,206 @@
...
+    public int hashCode() {
+        return Objects.hash(environmentName, environmentDisplayName, organizationName, solaceTopicsObject, publishTopics, subscribeTopics, solaceURLs);
+    }
+
+    @Override
+    public String toString() {
+        StringBuilder sb = new StringBuilder();
+        sb.append("class ApplicationSolaceDeployedEnvironmentsDTO {\n");
+
+        sb.append("    environmentName: ").append(toIndentedString(environmentName)).append("\n");
```
Comment (dnwick) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.store.v1/src/gen/java/org/wso2/carbon/apimgt/rest/api/store/v1/dto/ApplicationSolaceDeployedEnvironmentsDTO.java:185]: why we keep spaces here ?

**Review by dnwick — CHANGES_REQUESTED** [review]

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.admin.v1/src/gen/java/org/wso2/carbon/apimgt/rest/api/admin/v1/dto/AdditionalPropertyDTO.java (hunk 2 above)
Comment (Chamindu36) [reply components/apimgt/org.wso2.carbon.apimgt.rest.api.admin.v1/src/gen/java/org/wso2/carbon/apimgt/rest/api/admin/v1/dto/AdditionalPropertyDTO.java:13]: The DTO Files in the gen directory are auto-generated ones. Hence these things cannot be fixed there.

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1.common/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/common/mappings/APIMappingUtil.java [hunk 10]
```diff
@@ -1882,6 +1903,12 @@ public static AsyncAPISpecificationValidationResponseDTO getAsyncAPISpecificatio
                 infoDTO.setDescription(modelInfo.getDescription());
                 infoDTO.setEndpoints(modelInfo.getEndpoints());
                 infoDTO.setProtocol(model.getProtocol());
+                if (SolaceConstants.SOLACE_ENVIRONMENT.equals(new SolaceApiParser().getVendorFromExtension(model.
```
Comment (tharindu1st) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1.common/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/common/mappings/APIMappingUtil.java:1906]: do we need this solace related implementation?

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1.common/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/common/mappings/EnvironmentMappingUtil.java [hunk 11]
```diff
@@ -50,8 +61,44 @@ public static EnvironmentDTO fromEnvironmentToDTO(Environment environment) {
         environmentDTO.setType(environment.getType());
         environmentDTO.setServerUrl(environment.getServerURL());
         environmentDTO.setShowInApiConsole(environment.isShowInConsole());
+        environmentDTO.setProvider(environment.getProvider());
         environmentDTO.setVhosts(environment.getVhosts().stream().map(EnvironmentMappingUtil::fromVHostToVHostDTO)
                 .collect(Collectors.toList()));
+        environmentDTO.setAdditionalProperties(fromAdditionalPropertiesToAdditionalPropertiesDTO
+                (environment.getAdditionalProperties()));
+
+        if (SolaceConstants.SOLACE_ENVIRONMENT.equalsIgnoreCase(environment.getProvider())) {
```
Comment (tharindu1st) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1.common/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/common/mappings/EnvironmentMappingUtil.java:70]: do we need hardcode solace

**Review by tharindu1st — CHANGES_REQUESTED** [review]

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1.common/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/common/mappings/APIMappingUtil.java (hunk 10 above)
Comment (Chamindu36) [reply components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1.common/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/common/mappings/APIMappingUtil.java:1906]: Yes. To show the available endpoint URLs on the publisher portal for Solace deployments we need to use the Solace Admin APIs. Hence this part is needed.

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1.common/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/common/mappings/EnvironmentMappingUtil.java (hunk 11 above)
Comment (Chamindu36) [reply components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1.common/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/common/mappings/EnvironmentMappingUtil.java:70]: To show the available endpoint URLs on the publisher portal for Solace deployments we need to use the Solace Admin APIs. Hence this part is needed.

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/test/java/org/wso2/carbon/apimgt/impl/definitions/AsyncApiParserTest.java [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+import org.wso2.carbon.apimgt.api.model.APIIdentifier;
+
+import java.io.File;
+import java.util.HashMap;
+import java.util.Map;
+
+public class AsyncApiParserTest {
+
+    @Test
+    public void testGenerateAsyncAPIDefinition() throws Exception {
```
Comment (dnwick) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/test/java/org/wso2/carbon/apimgt/impl/definitions/AsyncApiParserTest.java:31]: @Chamindu36 Did we write test for failure scenarios as well ? Whether we throw the exceptions etc

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1.common/src/gen/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/dto/GatewayEnvironmentProtocolURIDTO.java [hunk 13]
```diff
@@ -0,0 +1,103 @@
...
+  @Override
+  public int hashCode() {
+    return Objects.hash(protocol, endpointURI);
+  }
+
+  @Override
+  public String toString() {
+    StringBuilder sb = new StringBuilder();
+    sb.append("class GatewayEnvironmentProtocolURIDTO {\n");
+
```
Comment (dnwick) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1.common/src/gen/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/dto/GatewayEnvironmentProtocolURIDTO.java:85]: remove new line

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/impl/ApisApiServiceImpl.java [hunk 14]
```diff
@@ -5057,6 +5059,12 @@ private APIDTO importAsyncAPISpecification(InputStream definition, String defini
             if (isServiceAPI) {
                 apiDTOFromProperties.setType(PublisherCommonUtils.getAPIType(service.getDefinitionType(), protocol));
             }
+            if (SolaceConstants.SOLACE_ENVIRONMENT.equals(apiDTOFromProperties.getGatewayVendor())) {
+                apiDTOFromProperties.getPolicies().add("AsyncUnlimited");
```
Comment (dnwick) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/impl/ApisApiServiceImpl.java:5063]: is it make sense add these as constants ?

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.store.v1/src/gen/java/org/wso2/carbon/apimgt/rest/api/store/v1/dto/AdditionalSubscriptionInfoListDTO.java [hunk 15]
```diff
@@ -0,0 +1,128 @@
...
+  @ApiModelProperty(example = "1", value = "Number of additional information sets of subscription returned. ")
+  @JsonProperty("count")
+  public Integer getCount() {
+    return count;
+  }
+  public void setCount(Integer count) {
+    this.count = count;
+  }
+
+  /**
```
Comment (dnwick) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.store.v1/src/gen/java/org/wso2/carbon/apimgt/rest/api/store/v1/dto/AdditionalSubscriptionInfoListDTO.java:49]: Dont leave empty comments

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.store.v1/src/gen/java/org/wso2/carbon/apimgt/rest/api/store/v1/dto/AdditionalSubscriptionInfoListDTO.java [hunk 16]
```diff
@@ -0,0 +1,128 @@
...
+import java.util.Objects;
+
+import javax.xml.bind.annotation.*;
+import org.wso2.carbon.apimgt.rest.api.common.annotations.Scope;
+import com.fasterxml.jackson.annotation.JsonCreator;
+
+import javax.validation.Valid;
+
+
+
```
Comment (dnwick) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.store.v1/src/gen/java/org/wso2/carbon/apimgt/rest/api/store/v1/dto/AdditionalSubscriptionInfoListDTO.java:24]: Add class comments

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.store.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/store/v1/mappings/AdditionalSubscriptionInfoMappingUtil.java [hunk 17]
```diff
@@ -0,0 +1,467 @@
...
+                                            }
+                                            solaceDeployedEnvironmentsDTO.setSolaceTopicsObject(solaceTopicsObjectDTO);
+                                        }
+                                    }
+                                }
+                                solaceEnvironments.add(solaceDeployedEnvironmentsDTO);
+                            }
+                        }
+                    } catch (IOException e) {
+                        e.printStackTrace();
```
Comment (dnwick) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.store.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/store/v1/mappings/AdditionalSubscriptionInfoMappingUtil.java:219]: Remove printStackTrace and add proper log with context. Do a full search and remove all please

**Review by dnwick — CHANGES_REQUESTED** [review]

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1.common/src/gen/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/dto/GatewayEnvironmentProtocolURIDTO.java (hunk 13 above)
Comment (Chamindu36) [reply components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1.common/src/gen/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/dto/GatewayEnvironmentProtocolURIDTO.java:85]: This file is an auto-generated DTO file. Hence we cannot change anything on this file.

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.store.v1/src/gen/java/org/wso2/carbon/apimgt/rest/api/store/v1/dto/AdditionalSubscriptionInfoListDTO.java (hunk 15 above)
Comment (Chamindu36) [reply components/apimgt/org.wso2.carbon.apimgt.rest.api.store.v1/src/gen/java/org/wso2/carbon/apimgt/rest/api/store/v1/dto/AdditionalSubscriptionInfoListDTO.java:49]: This file is an auto-generated DTO file. Hence we cannot change anything on this file.

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.store.v1/src/gen/java/org/wso2/carbon/apimgt/rest/api/store/v1/dto/AdditionalSubscriptionInfoListDTO.java (hunk 16 above)
Comment (Chamindu36) [reply components/apimgt/org.wso2.carbon.apimgt.rest.api.store.v1/src/gen/java/org/wso2/carbon/apimgt/rest/api/store/v1/dto/AdditionalSubscriptionInfoListDTO.java:24]: This file is an auto-generated DTO file. Hence we cannot change anything on this file.

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/test/java/org/wso2/carbon/apimgt/impl/definitions/AsyncApiParserTest.java (hunk 12 above)
Comment (Chamindu36) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/test/java/org/wso2/carbon/apimgt/impl/definitions/AsyncApiParserTest.java:32]: Fixed with https://github.com/wso2/carbon-apimgt/pull/10842/commits/c41ba333ee522fd29e341c130570450703f787bf

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/impl/ApisApiServiceImpl.java (hunk 14 above)
Comment (Chamindu36) [reply components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/impl/ApisApiServiceImpl.java]: Fixed with https://github.com/wso2/carbon-apimgt/pull/10842/commits/c41ba333ee522fd29e341c130570450703f787bf

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.store.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/store/v1/mappings/AdditionalSubscriptionInfoMappingUtil.java (hunk 17 above)
Comment (Chamindu36) [reply components/apimgt/org.wso2.carbon.apimgt.rest.api.store.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/store/v1/mappings/AdditionalSubscriptionInfoMappingUtil.java]: Fixed with https://github.com/wso2/carbon-apimgt/pull/10842/commits/c41ba333ee522fd29e341c130570450703f787bf

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/test/java/org/wso2/carbon/apimgt/impl/definitions/AsyncApiParserTest.java [hunk 18]
```diff
@@ -0,0 +1,98 @@
...
+        AsyncApiParser asyncApiParser = new AsyncApiParser();
+        try {
+            asyncApiParser.generateAsyncAPIDefinition(api);
+        } catch (JSONException e) {
+            Assert.assertNotNull(e);
+        }
+    }
+
+    @Test
+    public void testGetAsyncAPIDefinitionForStoreFail() throws Exception {
```
Comment (dnwick) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/test/java/org/wso2/carbon/apimgt/impl/definitions/AsyncApiParserTest.java:80]: can there be a case where async api will be generated for wrong format of ws url ?

Comment (Chamindu36) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/test/java/org/wso2/carbon/apimgt/impl/definitions/AsyncApiParserTest.java:80]: No. This is to create a mock scenario of WS URLs. There are checks to validate the URL when adding to the API object in actual flows.

**Review by dnwick — APPROVED** [review]

**Review by tharindu1st — CHANGES_REQUESTED** [review]

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1.common/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/common/mappings/EnvironmentMappingUtil.java [hunk 19]
```diff
@@ -20,15 +20,21 @@
 
 package org.wso2.carbon.apimgt.rest.api.publisher.v1.common.mappings;
 
+import org.wso2.carbon.apimgt.api.model.AsyncProtocolEndpoint;
 import org.wso2.carbon.apimgt.api.model.Environment;
 import org.wso2.carbon.apimgt.api.model.VHost;
+import org.wso2.carbon.apimgt.impl.ExternalEnvironment;
+import org.wso2.carbon.apimgt.impl.internal.ServiceReferenceHolder;
```
[the remark this answers is not in the document]
Comment (Chamindu36) [reply components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1.common/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/common/mappings/EnvironmentMappingUtil.java:27]: To get the registered ExternalEnvironment types, this was required.

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/pom.xml [hunk 20]
```diff
@@ -30,8 +30,24 @@
...
     <dependencies>
+        <dependency>
+            <groupId>com.fasterxml.jackson.core</groupId>
+            <artifactId>jackson-databind</artifactId>
+            <version>2.10.3</version>
+        </dependency>
+        <dependency>
+            <groupId>com.fasterxml.jackson.dataformat</groupId>
+            <artifactId>jackson-dataformat-yaml</artifactId>
+            <version>2.10.3</version>
```
[the remark this answers is not in the document]
Comment (Chamindu36) [reply components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/pom.xml]: Fixed with https://github.com/wso2/carbon-apimgt/pull/10842/commits/4c9042715f8f976ed78c923dae6754de7abd7f6af

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/pom.xml [hunk 21]
```diff
@@ -30,8 +30,24 @@
...
+            <id>jitpack.io</id>
+            <url>https://jitpack.io</url>
+        </repository>
+    </repositories>
 
     <dependencies>
+        <dependency>
+            <groupId>com.fasterxml.jackson.core</groupId>
+            <artifactId>jackson-databind</artifactId>
+            <version>2.10.3</version>
```
[the remark this answers is not in the document]
Comment (Chamindu36) [reply components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/pom.xml]: Fixed with https://github.com/wso2/carbon-apimgt/pull/10842/commits/4c9042715f8f976ed78c923dae6754de7abd7f6a

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/pom.xml [hunk 22]
```diff
@@ -30,8 +30,24 @@
     <name>WSO2 API Manager Publisher REST API - v2</name>
     <description>WSO2 API Manager REST API for API Publisher</description>
 
+    <repositories>
+        <repository>
+            <id>jitpack.io</id>
+            <url>https://jitpack.io</url>
```
[the remark this answers is not in the document]
Comment (Chamindu36) [reply components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/pom.xml]: Fixed with https://github.com/wso2/carbon-apimgt/pull/10842/commits/4c9042715f8f976ed78c923dae6754de7abd7f6a

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1.common/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/common/mappings/PublisherCommonUtils.java [hunk 23]
```diff
@@ -1139,8 +1145,9 @@ public static String updateAsyncAPIDefinition(String apiId, APIDefinitionValidat
 
         AsyncApiParser asyncApiParser = new AsyncApiParser();
         // Set uri templates
-        Set<URITemplate> uriTemplates = asyncApiParser.getURITemplates(
-                apiDefinition, APIConstants.API_TYPE_WS.equals(existingAPI.getType()));
+        Set<URITemplate> uriTemplates = asyncApiParser.getURITemplates(apiDefinition, APIConstants.
+                API_TYPE_WS.equals(existingAPI.getType()) || SolaceConstants.SOLACE_ENVIRONMENT.equals
```
[the remark this answers is not in the document]
Comment (Chamindu36) [reply components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1.common/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/common/mappings/PublisherCommonUtils.java]: Fixed with https://github.com/wso2/carbon-apimgt/pull/10842/commits/4c9042715f8f976ed78c923dae6754de7abd7f6a

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1.common/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/common/mappings/EnvironmentMappingUtil.java (hunk 19 above)
[the remark this answers is not in the document]
Comment (tharindu1st) [reply components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1.common/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/common/mappings/EnvironmentMappingUtil.java:27]: ServiceReferenceHolder can't access across packages due to this package is private from osgi.

**Review by tharindu1st — APPROVED** [review]

**Pull request merged**